### DISCUSSION TITLE
Adjoint solver with independent material parameters

### DISCUSSION THREAD

**Discussion by dritory** [NanoComp/meep#1973] [Q&A] [answered]
Hi!
The material grid interpolates between two materials of different dielectric properties, and the adjoint solver calculates the gradient with respect to the weights of the material grid. However, in my case I would like to express the adjoint solver optimization with respect to the permittivity _and_ the conductivity of the material grid, independently. This implies calculating two sensitivities; one for the permittivity and one for the conductivity, and optimizing over both.

As mentioned in #1747, it should be possible to map a function to the material grid and backpropagate the gradients, however, I don't know how this could be done in my case. Currently I have mapped the material grid directly onto the geometry of the simulation. Is there a way to solve this while still using the existing adjoint solver implementation?

**Comment by smartalecH** [comment, marked as answer]
> Is there a way to solve this while still using the existing adjoint solver implementation?

Yes and no. It depends on what you are *really* trying to do.

Computing the gradient w.r.t. ε and σ directly is not currently supported out of the box. There are a lot of fundamental issues with this idea (for example, only part of the tensors for ε and σ are stored/used at each point, thanks to the Yee grid). The `MaterialGrid` abstraction aims to simplify this issue.

Keep in mind that the underlying goal of the adjoint solver is density-based topology optimization, which means the user is continuously interpolating between two materials. These materials can, in fact, contain susceptibilities, conductivities, etc. But the idea is to interpolate between two *distinct* materials without introducing any field instabilities.

Based on your comment above, it sounds like you are interested in a different problem. It seems you don't care *what* the final materials look like. Rather, you just want to find the spatial permittivity and conductivity that maximize/minimize some objective function.

One way to accomplish this is the interpolate between vacuum and an "artificial material" with an upper bound for the conductivity and permittivity. The material grid will linearly interpolate between these two materials, but in a way to ensure each simulation remains stable. This means you will still compute gradients w.r.t. the `MaterialGrid` weights themselves.

There's more discussion behind this idea in our recent [Optics Express paper](https://doi.org/10.1364/OE.442074).

**Reply by dritory** [reply]
Thank you for your in-depth answer. Indeed what I am trying to do is not so related to density based topology optimization, but rather estimate the unknown permittivity and conductivity of the materials that minimize my objective function. 

The idea of interpolating between two materials which marks the lower and upper bound of the search domain should work for estimating a single material parameter. However, for two material parameters, like electric conductivity and permittivity, the interpolation would constrain the material on a fixed relation between the two parameters. What I am trying to do is calculating the gradient with respect to both material parameters, i.e. two weights or densities at each point. At least that is what I think I'm trying to do.

To me this seems impossible to achieve without modifications of the underlying C++ code. However, if there is any workaround please let me know. Perhaps something clever might be done with overlapping material grids.

I have an idea of how to solve this in a different way, by solving the parameters in two runs. By keeping one of the parameters constant while varying the other the gradient would be calculated with respect to the varied parameter. Though this would require one run for each parameter, which is unfortunate.

I must say, the rate and depth of the answers on the Meep project is admirable. Thanks!
